### PR TITLE
Coordinates package uses new iterator interfaces now.

### PR DESCRIPTION
**Pull request by marscher**
- TICA uses covartools (only one pass over data needed)
- Iterator states can be changed during iteration

There are still some things todo, but we want to reintegrate with devel now. All tests are passing.

Big thanks fly out to @clonker and @fabian-paul for their great work and valuable discussions!

### CODE REVIEW

**Comment by franknoe** [comment]
Oh my god. This is an enormous PR. I think we need to be careful with this merge (in addition to fixing the test failures):
- we need to do some benchmarks, showing that we haven't slowed down operations before merging this. 
- I'd like to see if all notebooks still give the same results.

**Comment by franknoe** [comment]
Also, if this already introduces the new iterator interface, we should meet and discuss it first, because it's a design decision. Tomorrow afternoon?

**Comment by marscher** [comment]
I have opened up this PR, since all tests are now passing (an we really made sure that we do not break any api or program behaviour).

The tests of this branch succeeded on my fork, but failed on the central repo because of this:
https://travis-ci.org/marscher/PyEMMA/builds/101880721

```
Error: HTTPError: 404 Client Error: Not Found for url: https://conda.anaconda.org/omnia/linux-64/msmtools-1.1-np19py27_2.tar.bz2: https://conda.anaconda.org/omnia/linux-64/msmtools-1.1-np19py27_2.tar.bz2 
```

The design is mainly what we have discussed during our last meeting. Of course we can meet tomorrow whenever you tight schedule allows.

**Comment by franknoe** [comment]
Am 12/01/16 um 22:26 schrieb Martin K. Scherer:

> I have opened up this PR, since all tests are now passing (an we 
> really made sure that we do not break any api or program behaviour).
> 
> Yep, but that doesn't prove that the notebooks still have the same 
> behavior, and that the performance is either similar or better. And with 
> such a huge change I'd like to see that (we can also discuss if we can 
> automatize that to some degree).
> 
> The tests of this branch succeeded on my fork, but failed on the 
> central repo because of this:
> https://travis-ci.org/marscher/PyEMMA/builds/101880721
> 
> |Error: HTTPError: 404 Client Error: Not Found for url: 
> https://conda.anaconda.org/omnia/linux-64/msmtools-1.1-np19py27_2.tar.bz2: 
> https://conda.anaconda.org/omnia/linux-64/msmtools-1.1-np19py27_2.tar.bz2 
> ``` The design is mainly what we have discussed during our last 
> meeting. Of course we can meet tomorrow whenever you tight schedule 
> allows.|
> |We discussed a design proposal, but we never met to review the actual 
> design. Tomorrow afternoon is fine.|
> ||
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/pull/671#issuecomment-171061605.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by marscher** [comment]
We'll add the results of these integration and performance tests to this PR, to track the issues, tackle em down and finally integrate this beast into devel.

I'm still looking for a best practice to run performance regression tests on some CI. Travis is obviously not suited for that, since one gets a different system all the time. But I will try to integrate this into our Jenkins service.
Same should be applied for long running notebook tests.

File: pyemma/coordinates/clustering/uniform_time.py [hunk 1]
```diff
@@ -30,94 +33,66 @@
...
+        # first data point in the middle of the time segment
+        next_t = (T // self.n_clusters) // 2
+        # cumsum of lenghts
+        cumsum = np.cumsum(self.trajectory_lengths())
+        # distribution of integers, truncate if n_clusters is too large
+        linspace = self.stride * np.arange(next_t, T - next_t + 1, (T - 2*next_t + 1) // self.n_clusters)[:self.n_clusters]
+        # random access matrix
+        ra_stride = np.array([UniformTimeClustering._idx_to_traj_idx(x, cumsum) for x in linspace])
+        with iterable.iterator(stride=ra_stride, return_trajindex=False) as it:
+            assert it is not None
```
Comment (marscher) [inline pyemma/coordinates/clustering/uniform_time.py:85]: @marscher remove this.

File: pyemma/coordinates/data/feature_reader.py [hunk 2]
```diff
@@ -173,36 +181,155 @@ def dimension(self):
...
+        return data
+
+
+class FeatureReaderJaggedRandomAccessStrategy(FeatureReaderCuboidRandomAccessStrategy):
+    def _get_itraj_random_accessible(self, itrajs, frames, dims):
+        itrajs = self._get_indices(itrajs, self._source.ntraj)
+        return [self._source._ra_cuboid[itraj, frames, dims][0] for itraj in itrajs]
+
+
+class FeatureReaderLinearItrajRandomAccessStrategyFeatureReader(FeatureReaderCuboidRandomAccessStrategy):
```
Comment (marscher) [inline pyemma/coordinates/data/feature_reader.py:250]: featurereader featurereaderreadereaderfeature

File: pyemma/coordinates/data/writer.py [hunk 3]
```diff
@@ -26,27 +26,25 @@
...
-from pyemma.coordinates.transform.transformer import Transformer
+from pyemma.coordinates.transform.transformer import StreamingTransformer
 
 
-class WriterCSV(Transformer):
+class WriterCSV(StreamingTransformer):
 
     '''
-    shall write to csv files
+    write to csv files via numpy
```
Comment (marscher) [inline pyemma/coordinates/data/writer.py:35]: move this stuff to a public method in DataSource. Eg. "write_to_csv(filename,...)"

File: pyemma/coordinates/util/patches.py [hunk 4]
```diff
@@ -33,111 +32,173 @@
...
-            yield t[i:i+chunk]
-
-    elif isinstance(stride, np.ndarray):
-        with (lambda x: open(x, n_atoms=topology.n_atoms)
-              if extension in ('.crd', '.mdcrd')
-              else open(filename))(filename) as f:
-            x_prev = 0
+
+
+class iterload:
```
Comment (marscher) [inline pyemma/coordinates/util/patches.py:37]: new style class

**Comment by franknoe** [comment]
I really like the new code. Thanks a lot to  @marscher , @clonker  and @fabian-paul for assembling this! We need to intensively test this next week.

**Pull request merged**